### ISSUE TITLE
Why not merge as extension in IdeaVim repo?

### ISSUE THREAD

**Issue by jannes** [Mishkun/ideavim-sneak#4]
Hey, 
what is the reasoning that you made this a separate plugin as opposed to merging it to the IdeaVim repo?
I would personally prefer it that way, but interested to hear your thoughts:)

**Comment by Mishkun** [comment]
For me it was a rather cool project to hack up in two evenings "just for myself" and then a friend of mine told me that I could also share it to the public. 

I don't really know what is the vision of the core team for the IdeaVim – is it more like a monorepo thing trying to reimplement all of the most popular plugins or it is more like a foundation which can help others to build their favorite plugins by themselves. Even if former is true – how could I know if `vim-sneak` is good enough to be part of the repo?

I would be glad if my extension would become part of the IdeaVim repo, It would be easier to maintain it this way. I'll get in touch with maintainers to discuss this

**Comment by jannes** [comment]
Ah okay I see.
Yea it is a single repository that already has a few vim plugins incorporated in the 'extension' folder.  
(https://github.com/JetBrains/ideavim/tree/master/src/com/maddyhome/idea/vim/extension)  

I think in order to add this project to the repo we should make the jumps vertical like mentioned in the other issue  
and also implement the repeat actions in a less hacky way (currently this overrides all repeat behavior except the ones   
for 'f' and 't' which you just forward, so for example things like 'w', 'b' are not repeatable anymore).

I would be happy to help, if you make it a goal to get this incorporated into the main repo:)

**Comment by Mishkun** [comment]
Yeah, it should be a good thing to merge this plugin into IdeaVim repo. It would be easier to implement a proper repeat behaviour using `MotionGroup` to store actions

**Comment by AlexPl292** [comment]
Hi folks!

> I don't really know what is the vision of the core team for the IdeaVim – is it more like a monorepo thing trying to reimplement all of the most popular plugins or it is more like a foundation which can help others to build their favorite plugins by themselves. 

To be honest, initially we didn't have any plans to build a monorepo or foundation. Some plugins were just implemented along with IdeaVim. Technically, it is now a monorepo. The fact that side IJ plugins can contribute extensions to IdeaVim is more a side effect of the IdeaVim extensions architecture than a planned solution.

@Mishkun I've checked the overall implementation of the plugin and it looks good for me. Since you're okay with the idea of merging, we can do it that way. That's would simplify the flow for the users (no additional plugins) and the extension would be under maintenance as long as IdeaVim is.
AFAIU @jannes has made a more interesting investigation and found some issues, but I think there'll be no problem to merge this work into the main repo after the fixes. For now, my proposal is to continue the work here, and after that, I'll think on how to transfer the code in the IdeaVim repo in the best way (hope to make it with keeping authorship of the commits, it's not the least thing after all 🙂).

As usual, I'm here for you if you'll need some help with developing. For example, I don't know at the moment how to detect visible lines in the editor, but I can try to find it quickly if you'll need. As for storing actions in `MotionGroup`, I don't know what exactly you mean at the moment, but if it's extension-specific information, it would be better to store it somewhere inside the extensions code. `MotionGroup` and other core classes are made to reuse the existing code between different vim actions.

**Comment by Mishkun** [comment]
@AlexPl292 Thanks for your kind words. I'll try my best to implement this extension in a nicer way.

`MotionGroup` remembers the last f/F/t/T motion in `lastFTCmd` variable for repeating using `repeatLastMatchChar`. This is why I had to workaround this by intercepting ;/, keystrokes. Maybe it would better to have some listener/holder to store latest movement. It would help with repeating any other kind of motion.

Thanks for your help, again.

**Comment by UnsolvedCypher** [comment]
@AlexPl292 and @Mishkun - how would you feel about doing the merge relatively soon? I understand that there are some improvements that could be made to the implementation. However, the extension is very usable as-is without any bugs, and is well-tested,  so I think other IdeaVim users could benefit from it. Currently, it's not listed under the supported plugins section so it's hard for users to find. 

Additionally, putting it in the main repo would give it more attention from IdeaVim contributors, so missing features could be added faster while taking some of the weight off of @Mishkun. What do you think?

**Comment by Mishkun** [comment]
@UnsolvedCypher I think that this is a good idea to merge it in. It will provide more tight integration with IdeaVim and allow some corners to be cut. I will prepare a PR to IdeaVim soon

**Comment by AlexPl292** [comment]
I don't know what happened, but I completely missed that a YEAR has passed since we have discussed that possibility. Looks like vimscript support took to much of my attention. 😅
The PR would be highly appreciated. Hope the review won't take that much :D

**Comment by UnsolvedCypher** [comment]
Hey @Mishkun, are you still planning to submit a PR for this? If there is anything I can do to help with the process, please let me know!  Thanks again for your plugin, I use it every day 😃
